### ISSUE TITLE
ClassCastException in MvpNullObjectBasePresenter

### ISSUE THREAD

**Issue by manabreak** [sockeqwe/mosby#194]
I ran into this problem when using MvpNullObjectBasePresenter. When a fragment is destroyed, the following error crashes the application:

    java.lang.ClassCastException: java.lang.Class cannot be cast to java.lang.reflect.ParameterizedType
        at com.hannesdorfmann.mosby.mvp.MvpNullObjectBasePresenter.detachView(MvpNullObjectBasePresenter.java:38)

Took a quick glance at the source, but didn't see a solution right away. I thought it was because my views used two generic types for the presenter (`P extends MvpPresenter<V> & Refreshable`), but after taking the other type out, the problem persists.

The problem happens when this fragment gets destroyed:

    public class InboxFragment extends MessageBaseFragment<InboxPresenter> {
        @NonNull
        @Override
        public InboxPresenter createPresenter() {
            return new InboxPresenter(...);
        }
    }

`MessageBaseFragment` looks like this (unrelevant code omitted):

    public abstract class MessageBaseFragment<P extends BaseMessagePresenter> extends BaseMvpFragment<MessageView, P> implements MessageView {
        ...
    }

And the `BaseMvpFragment` is just a small container for injected dependencies:

    public abstract class BaseMvpFragment<V extends MvpView, P extends MvpPresenter<V>> extends MvpFragment<V, P> {
        ...
    }

Okay, moving on, the `InboxPresenter`:

    public class InboxPresenter extends BaseMessagePresenter {
        ...
    }

And `BaseMessagePresenter`:

    public abstract class BaseMessagePresenter extends MvpNullObjectBasePresenter<MessageView> {
        ...
    }

Finally, `MessageView`:

    public interface MessageView extends MvpView {
        ...
    }

**Comment by sockeqwe** [comment]
Which version of Mosby are you using?

manabreak <notifications@github.com> schrieb am Sa., 21. Jan. 2017, 12:09:

**Comment by manabreak** [comment]
I'm using 2.0.1.

**Comment by sockeqwe** [comment]
This should be fixed with 3.0.0 snapshot. Can you try if this is fixed
with  3.0 snapshot?

manabreak <notifications@github.com> schrieb am Sa., 21. Jan. 2017, 12:45:

**Comment by manabreak** [comment]
Hmm, Gradle seems to be having problems resolving `compile 'com.hannesdorfmann.mosby3:mvp:3.0.0-SNAPSHOT'`. Is there something that should be changed in addition?

**Comment by sockeqwe** [comment]
You also have to add the url to the snapshot repository:

```gradle
allprojects {
  repositories {
    ...

    maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
}
```

**Comment by sockeqwe** [comment]
And the package  name of all classes has been changed from com.hannesdorfmann.mosby to com.hannesdorfmann.mosby**3**

Simply do a search and replace in Android Studio to change the package name in your classes import statements.

**Comment by manabreak** [comment]
The issue seems to be solved when using 3.0.0 snapshot. Thanks!

**Comment by cyberrob** [comment]
Comparing com.hannesdorfmann.mosby.mvp.MvpNullObjectBasePresenter & com.hannesdorfmann.mosby3.mvp.MvpNullObjectBasePresenter, it seems that constructor in the newer version did some magics to recursively check if subclasses is some kind of class in __isSubTypeOfMvpView()__ method.

It's awesome! Thanks!
